### DISCUSSION TITLE
Korge 3D

### DISCUSSION THREAD

**Discussion by dhakehurst** [korlibs/korge#2267] [Tutorials]
Hi,
I'm trying to update to latest version of korge (from old version 2 ish)

Is there still support for 3D, skyboxes etc.
I can't find any up-to-date 3D samples.

thanks

**Comment by soywiz** [comment]
Hi David!

Since 3D was still evolving and was not mature enough, we moved it to a separate repo so it could be evolved at a different pace, with the idea to eventually merge it again once mature enough.

Here's the repo: https://github.com/korlibs/korge-k3d

**Comment by dhakehurst** [comment]
Hi, thanks,
I can't find it on maven central?
would someone publish it.
thanks

**Comment by dhakehurst** [comment]
Or maybe its easier if I check it out and build locally, though would be nice it it got published ;-) 

(where would be best to discus build issues? )

**Comment by rafi0** [comment]
@dhakehurst hi, i also wanted to update to latest version and use some 3d, but the issue is, that some code is commented out there:
terrain: https://github.com/korlibs/korge-k3d/blob/main/korge-k3d/src/commonMain/kotlin/korlibs/korge3d/Terrain3D.kt
skybox: https://github.com/korlibs/korge-k3d/blob/main/korge-k3d/src/commonMain/kotlin/korlibs/korge3d/SkyBox.kt
lights and much more

i tried uncommenting and getting it to compile, but it needs alot of time to get it done, so i gave up after an hour.
I might try again in some time. I encourage to try that also if you are interested. We might even make a call about this if you are able to help with that.

**Reply by dhakehurst** [reply]
Hi,
Glad to hear someone else is interested. ;-) 
I made the original PR to add the skybox (terrain as well maybe, I forget)

I was very happy it got accepted, and now disappointed that it has not been maintained, in sufficient interest by the owner and other users I guess.


I will have a look and see if I can get it to work again, but a lot has changed since I last worked on the code base.
I'll let you know how I get on.

**Comment by dhakehurst** [comment]
Currently I can't get it to run, I get this error
```
java.lang.reflect.InaccessibleObjectException: Unable to make public static boolean sun.java2d.opengl.OGLUtilities.invokeWithOGLContextCurrent(java.awt.Graphics,java.lang.Runnable) accessible: module java.desktop does not "opens sun.java2d.opengl" to unnamed module @78a331fe
....
Might require run the JVM with --add-opens=java.desktop/sun.java2d.opengl=ALL-UNNAMED
....
```

I've tried may options but can't get past this error.

any suggestions ?

thanks

**Comment by soywiz** [comment]
BTW. `korge-k3d` is supported. It is just that it was moved from the core to outside so it could be evolved separately. It was evolved to support skinning, GLTF among other stuff. Not my priority right now since I'm working on the core and forge; so if you could evolve it that would be awesome.

The way to use it outside the project is via kproject: https://store.korge.org/module/korge-k3d/
Documentation of kproject: https://docs.korge.org/kproject/

```bash
git clone git@github.com:korlibs/korge-k3d.git
cd korge-k3d
./gradlew runJvm
```

should yield this result:

<img width="1433" alt="Screenshot 2024-07-09 at 15 20 44" src="https://github.com/korlibs/korge/assets/570848/cf1739fa-ec91-44ec-8500-1ce6e890ce71">

If not, which JVM are you using and which OS?

**Reply by soywiz** [reply]
3D demo is also showcased here: https://show.korge.org/#K3DScene

**Comment by dhakehurst** [comment]
Thanks.
I am running on Mac (Apple M1 Pro)
With Temurin JVM  - 19.0.2 (Eclipse Adoptium 19.0.2+7)

Running from the command line does work.

but not from within IntelliJ (which makes debugging tricky!)
I guess there are some JVM args I need to configure in the IntelliJ Run Configurations,
struggling to find out what they are!

**Reply by dhakehurst** [reply]
though there do seem to be runtime errors running from command line

![image](https://github.com/korlibs/korge/assets/5618533/6516744b-7341-489f-9f44-300308df31e5)

**Reply by soywiz** [reply]
How are you calling it from IntelliJ? In theory you should run runJvm task in debug mode, and everything should work
